crates/nu-command/src/filters/insert.rs

```diff
@@ -118,97 +117,270 @@ fn insert(
+                                span,
+                                engine_state,
+                                &mut stack,
+                                redirect_stdout,
+                                redirect_stderr,
+                                block,
+                                &cell_path.members,
+                                false,
+                            )?;
                         }
                     }
+                    _ => {
+                        insert_single_value_by_closure(
+                            &mut value,
+                            replacement,
+                            engine_state,
+                            stack,
+                            redirect_stdout,
+                            redirect_stderr,
+                            &cell_path.members,
+                            false,
+                        )?;
+                    }
+                }
+            } else {
+                value.insert_data_at_cell_path(&cell_path.members, replacement, span)?;
+            }
+            Ok(value.into_pipeline_data_with_metadata(metadata))
+        }
+        PipelineData::ListStream(mut stream, metadata) => {
+            if let Some((
+                &PathMember::Int {
+                    val,
+                    span: path_span,
+                    ..
+                },
+                path,
+            )) = cell_path.members.split_first()
+            {
+                let mut pre_elems = vec![];
 
-                    let output = eval_block(
-                        &engine_state,
-                        &mut stack,
-                        &block,
-                        input.clone().into_pipeline_data(),
-                        redirect_stdout,
-                        redirect_stderr,
-                    );
-
-                    match output {
-                        Ok(pd) => {
-                            let span = pd.span().unwrap_or(span);
-                            if let Err(e) = input.insert_data_at_cell_path(
-                                &cell_path.members,
-                                pd.into_value(span),
-                                span,
-                            ) {
-                                return Value::error(e, span);
+                for idx in 0..val {
+                    if let Some(v) = stream.next() {
+                        pre_elems.push(v);
+                    } else {
+                        return Err(ShellError::InsertAfterNextFreeIndex {
+                            available_idx: idx,
+                            span: path_span,
+                        });
+                    }
+                }
+
+                if path.is_empty() {
+                    if replacement.as_block().is_ok() {
+                        let value = stream.next();
+                        let end_of_stream = value.is_none();
+                        let value = value.unwrap_or(Value::nothing(path_span));
```

Comment:
Only spot where I'm unsure what span to use. Use `[0 1] | every 1 | insert 2.0 value` to see for yourself.

Reply:
It seems like the spans should look like this in both examples. However, I'm not sure what "value not a list" is supposed to mean here.
```
Error: nu::shell::not_a_list

  × Not a list value
   ╭─[entry #7:1:1]
 1 │ [0 1] | insert 2.0 value
   · ──┬──          ─┬─
   ·   │             ╰── value not a list
   ·   ╰── value originates here
   ╰────
```